### PR TITLE
Fix completions that got broken in 1.3.0rc

### LINKED ISSUE #2393: 1.3.0.RC1: CLI tab complete broken
```
$ kontena ser/Users/lauri/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/kontena-cli-1.3.0.rc1/lib/kontena/cli/common.rb:14:in `<module:Common>': uninitialized constant Kontena::Cli::Spinner (NameError)
    from /Users/lauri/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/kontena-cli-1.3.0.rc1/lib/kontena/cli/common.rb:5:in `<module:Cli>'
    from /Users/lauri/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/kontena-cli-1.3.0.rc1/lib/kontena/cli/common.rb:4:in `<module:Kontena>'
    from /Users/lauri/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/kontena-cli-1.3.0.rc1/lib/kontena/cli/common.rb:3:in `<top (required)>'
    from /Users/lauri/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
    from /Users/lauri/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
    from /Users/lauri/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/kontena-cli-1.3.0.rc1/lib/kontena/scripts/completer.rb:2:in `<top (required)>'
    from /Users/lauri/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
    from /Users/lauri/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
    from /Users/lauri/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/kontena-cli-1.3.0.rc1/bin/kontena:13:in `<top (required)>'
    from /Users/lauri/.rbenv/versions/2.3.3/bin/kontena:22:in `load'
    from /Users/lauri/.rbenv/versions/2.3.3/bin/kontena:22:in `<main>'
```

### PR DESCRIPTION
**Pull request by kke**
Fixes #2393

### CODE REVIEW

**Comment by jnummelin** [comment]
I guess this is somewhat WIP still, at least the specs are failing hard.

Also, as we're fixing a bug, there definitely needs to be some specs for this.

**Comment by kke** [comment]
What was  failing if there are no specs? :) 

Bug was in common, enhancement was made to completer.

**Comment by jnummelin** [comment]
> What was failing if there are no specs? :)

https://travis-ci.org/kontena/kontena/jobs/238227080

From that amount of spec fails I assumed something got broke with some commits in this PR :)

Some speccing for the completion should be done. Would it be possible to do a spec that ensures that the completions do really load properly?

**Comment by kke** [comment]
e2e specs added and they're passing.

**Comment by jnummelin** [comment]
```
$ kontena grid use /Users/jussi/code/kontena/cli/lib/kontena/cli/config.rb:318:in `require_current_master_token': stack level too deep (SystemStackError)
	from /Users/jussi/code/kontena/cli/lib/kontena/cli/common.rb:152:in `require_token'
	from /Users/jussi/code/kontena/cli/lib/kontena/scripts/completer.rb:8:in `client'
	from /Users/jussi/code/kontena/cli/lib/kontena/cli/common.rb:159:in `rescue in require_token'
	from /Users/jussi/code/kontena/cli/lib/kontena/cli/common.rb:151:in `require_token'
	from /Users/jussi/code/kontena/cli/lib/kontena/scripts/completer.rb:8:in `client'
	from /Users/jussi/code/kontena/cli/lib/kontena/cli/common.rb:159:in `rescue in require_token'
	from /Users/jussi/code/kontena/cli/lib/kontena/cli/common.rb:151:in `require_token'
	from /Users/jussi/code/kontena/cli/lib/kontena/scripts/completer.rb:8:in `client'
	 ... 11218 levels...
	from /Users/jussi/.rvm/gems/ruby-2.3.3/bin/kontena:22:in `load'
	from /Users/jussi/.rvm/gems/ruby-2.3.3/bin/kontena:22:in `<main>'
	from /Users/jussi/.rvm/gems/ruby-2.3.3/bin/ruby_executable_hooks:15:in `eval'
	from /Users/jussi/.rvm/gems/ruby-2.3.3/bin/ruby_executable_hooks:15:in `<main>'
```

That's when I do `kontena grid use [tab]`. Somehow seems to go into never-ending recursion

File: test/spec/features/completer/completer_spec.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  it 'can complete subsubsubsubcommands' do
+    k = run 'kontena complete kontena master user role'
+    rows = k.out.split(/[\r\n]+/)
+    expect(rows).to include "add"
+    expect(rows).to include "remove"
+  end
+
+  it 'can complete master names' do
+    k = run 'kontena complete kontena master use'
+    expect(k.out).to match(/e2e/)
```
Comment (jnummelin) [inline test/spec/features/completer/completer_spec.rb]: It's not 100% guaranteed that there is a master with that name. I'm executing the e2e specs against random masters at least.

File: test/spec/features/completer/completer_spec.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  end
+
+  context 'master queries' do
+    context 'with current master set' do
+      it 'can complete node names' do
+        k = run 'kontena complete kontena node show'
+        expect(k.out).to match(/moby/)
```
Comment (jnummelin) [inline test/spec/features/completer/completer_spec.rb]: I don't think it's good idea to "hardcode" node names in these specs. IMO these should be easily runnable agains any setup. I bet @SpComb is not running these with any nodes named `moby`. :) 

Maybe capture known node list from `kontena node ls -q` and compare the completion list to that?

Yes, I know there's already some "expectations" on the setup, but I'm hoping we get rid of them anyway

File: test/spec/features/completer/completer_spec.rb [hunk 3]
```diff
@@ -51,14 +51,19 @@
 
   context 'master queries' do
     context 'with current master set' do
-      it 'can complete node names' do
+      it 'can complete grid names' do
```
Comment (jnummelin) [inline test/spec/features/completer/completer_spec.rb]: `it 'can complete node names' do ...` :)

File: test/spec/features/completer/completer_spec.rb [hunk 4]
```diff
@@ -51,14 +51,19 @@
...
+        node_names = run('kontena node ls -q').out.split(/[\r\n]+/)
```
Comment (jnummelin) [inline test/spec/features/completer/completer_spec.rb]: Needs some stripping:
```
Failure/Error: expect(rows).to include node_name
       expected ["core-01", "moby"] to include "moby   "
```

Comment (kke) [inline test/spec/features/completer/completer_spec.rb]: yes, looks like tty-table adds trailing space to single column output. e2e finding bugs like crazy.

File: cli/lib/kontena/scripts/completer.rb [hunk 5]
```diff
@@ -160,6 +188,8 @@ def subcommand_tree(cmd = nil, base = nil)
         completion.clear
         sub_commands = %w(list use user current remove rm config cfg login logout token join audit-log init-cloud)
         if words[1] && words[1] == 'use'
+          completion.push helper.master_names.reject { |n| n == helper.current_master_name }
```
Comment (kke) [inline cli/lib/kontena/scripts/completer.rb]: ^("master use" and "grid use" completions should not include the current master/grid)

**Comment by kke** [comment]
I think i'm finished with the pushes for now.

**Comment by jnummelin** [comment]
Tested this all the ways I could think of, e2e cases pass also nicely.

**Pull request merged**